File: frontend/src/api/media.js
/**
 * Media API — file upload and media library.
 *
 * Backend prefix: /api/media
 */

import { api } from './client.js';
import { captureVideoPoster, isVideoFile } from '../utils/videoPoster.js';

/**
 * List media items.
 *
 * `paths` switches the endpoint out of listing mode: it resolves exactly the
 * given content paths ("/YYYY/MM/file") and ignores the paging keys, which is
 * how a caller that already knows which media it wants — the post editor, say —
 * avoids fishing for them in a page of the library. At most 500 per request.
 *
 * `orphaned_only` is deliberately absent: the handler reads page, per_page,
 * file_type, folder and paths and nothing else, so any other key is sent and
 * dropped.
 *
 * @param {{ page?: number, per_page?: number, file_type?: string, folder?: string, paths?: string[] }} [params]
 * @returns {Promise<{ media: object[], total, page, per_page, pages }>}
 */
export function listMedia(params = {}) {
  return api.get('/api/media', params);
}

// Paths ride in the query string, and a photo essay can reference more of them
// than one URL should carry, so a lookup goes out in batches of this size.
const MEDIA_PATH_BATCH = 100;

/**
 * Resolve the media records at the given content paths, keyed by path.
 *
 * This is the lookup a post editor wants: a post's images are what its content
 * references, which is neither what `media.post_id` records nor what any one
 * page of the library happens to contain.
 *
 * @param {string[]} paths  Content paths, e.g. "/2026/03/1712345678_shot.jpg"
 * @returns {Promise<Record<string, object>>}
 */
export async function getMediaByPaths(paths) {
  const unique = [...new Set(paths.filter(Boolean))];
  const batches = [];
  for (let i = 0; i < unique.length; i += MEDIA_PATH_BATCH) {
    batches.push(listMedia({ paths: unique.slice(i, i + MEDIA_PATH_BATCH) }));
  }
  /** @type {Record<string, object>} */
  const byPath = {};
  for (const result of await Promise.all(batches)) {
    for (const m of result.media || []) if (m.path) byPath[m.path] = m;
  }
  return byPath;
}

/**
 * Get distinct year/month folders from the media library.
 * @param {{ file_type? }} [params]
 * @returns {Promise<{ folders: { year, month, path }[] }>}
 */
export function getMediaFolders(params = {}) {
  return api.get('/api/media/folders', params);
}

/**
 * Get a single media item by ID.
 * @param {number} id
 * @returns {Promise<object>}
 */
export function getMedia(id) {
  return api.get(`/api/media/${id}`);
}

/**
 * Upload a single file.
 *
 * A video is accompanied by a poster frame captured here in the browser — the
 * server cannot decode video, so this is the only chance to give the file a
 * thumbnail. Capture failures are silent and leave the video poster-less.
 *
 * @param {File}    file
 * @param {{ alt_text?, caption?, post_id? }} [meta]
 * @returns {Promise<object>}
 */
export async function uploadMedia(file, meta = {}) {
  const form = new FormData();
  form.append('file', file);
  if (meta.alt_text) form.append('alt_text', meta.alt_text);
  if (meta.caption)  form.append('caption', meta.caption);
  if (meta.post_id)  form.append('post_id', String(meta.post_id));

  if (isVideoFile(file)) {
    const poster = await captureVideoPoster(file);
    if (poster) form.append('poster', poster, 'poster.jpg');
  }

  return api.upload('/api/media/upload', form);
}

/**
 * Store a poster frame for an existing video, backfilling one that was
 * uploaded before posters existed or ingested outside the admin UI.
 * @param {number} id
 * @param {Blob}   poster  JPEG frame
 * @returns {Promise<object>} Updated media object
 */
export function setVideoPoster(id, poster) {
  const form = new FormData();
  form.append('poster', poster, 'poster.jpg');
  return api.upload(`/api/media/${id}/poster`, form);
}

/**
 * Upload multiple files.
 * @param {File[]} files
 * @param {number} [postId]
 * @returns {Promise<{ uploaded: object[], failed: object[], total_uploaded, total_failed }>}
 */
export function uploadMultiple(files, postId) {
  const form = new FormData();
  files.forEach((f) => form.append('files', f));
  if (postId) form.append('post_id', String(postId));
  return api.upload('/api/media/upload/multiple', form);
}

/**
 * Update media metadata (alt_text, caption, post_id, metadata).
 *
 * `metadata` is the EXIF map, replaced wholesale — the visual editor's per-image
 * EXIF panel saves through here (UpdateMediaRequest in api/internal/api/media.go
 * has always read it).
 *
 * @param {number} id
 * @param {{ alt_text?: string, caption?: string, post_id?: number,
 *           metadata?: Record<string, any> }} data
 * @returns {Promise<object>}
 */
export function updateMedia(id, data) {
  return api.patch(`/api/media/${id}`, data);
}

/**
 * Rename a media item.
 * @param {number} id
 * @param {string} newFilename
 * @returns {Promise<object>}
 */
export function renameMedia(id, newFilename) {
  return api.post(`/api/media/${id}/rename`, { new_filename: newFilename });
}

/**
 * Delete a media item.
 * @param {number} id
 * @returns {Promise<object>}
 */
export function deleteMedia(id) {
  return api.delete(`/api/media/${id}`);
}

/**
 * Get storage statistics.
 * @returns {Promise<object>}
 */
export function getMediaStats() {
  return api.get('/api/media/stats');
}

/**
 * List orphaned media files.
 * @returns {Promise<{ media: object[], total, total_size_bytes }>}
 */
export function getOrphanedMedia() {
  return api.get('/api/media/orphaned');
}

/**
 * Delete all orphaned media files.
 * @returns {Promise<object>}
 */
export function deleteOrphanedMedia() {
  return api.delete('/api/media/orphaned');
}

/**
 * Analyze an existing media item with AI (Gemini) to suggest title, tags, and excerpt.
 * @param {number} id
 * @returns {Promise<{ title: string|null, tags: string[], excerpt: string|null }>}
 */
export function analyzeMedia(id) {
  return api.post(`/api/media/${id}/analyze`);
}

/**
 * Analyze a stored media file by its URL path (e.g. "/2024/08/photo.jpg").
 * @param {string} path
 * @returns {Promise<{ title: string|null, tags: string[], excerpt: string|null }>}
 */
export function analyzeMediaByPath(path) {
  return api.post('/api/media/analyze-path', { path });
}

/**
 * Re-extract EXIF data from the original file on disk.
 * Overwrites any manually edited EXIF with camera-extracted values.
 * @param {number} id
 * @returns {Promise<object>} Updated media object
 */
export function reextractMediaEXIF(id) {
  return api.post(`/api/media/${id}/reextract`, {});
}

/**
 * Write EXIF fields back to the media file and update the DB.
 * Only alphanumeric and space characters are accepted.
 * @param {number} id
 * @param {Record<string, string>} fields  e.g. { Make: "Canon", Model: "EOS R5" }
 * @returns {Promise<object>} Updated media object
 */
export function updateMediaEXIF(id, fields) {
  return api.put(`/api/media/${id}/exif`, fields);
}

/**
 * Revert media EXIF metadata to the original values captured at upload.
 * @param {number} id
 * @returns {Promise<object>} Updated media object
 */
export function revertMediaEXIF(id) {
  return api.post(`/api/media/${id}/revert-exif`, {});
}

/**
 * Invalidate every derived image: the server purges the variant tree, rolls the
 * thumbnail generation token and regenerates the most recent uploads in the
 * background. There is nothing to opt out of — a rebuild discards every file.
 * @returns {Promise<{ message: string, stats: { generation: string, purged: number, legacy: number, prewarming: number } }>}
 */
export function rebuildThumbnails() {
  return api.post("/api/media/thumbnails/rebuild");
}

File: api/internal/api/media.go
package api

import (
	"errors"
	"fmt"
	"io"
	"math"
	"net/http"
	"regexp"
	"strconv"
	"strings"

	"point-api/internal/models"
	"point-api/internal/services"

	"github.com/labstack/echo/v4"
)

// maxPosterBytes bounds a video poster frame. Posters are downscaled to the
// thumbnail box on arrival, so anything larger than this is not a frame capture.
const maxPosterBytes = 8 << 20 // 8 MB

type MediaHandler struct {
	mediaService    *services.MediaService
	settingsService *services.SettingsService
}

func NewMediaHandler(mediaService *services.MediaService, settingsService *services.SettingsService) *MediaHandler {
	return &MediaHandler{
		mediaService:    mediaService,
		settingsService: settingsService,
	}
}

// mediaResponse is mediaToResponse with the request's thumbnail generation
// token already read off the settings cache.
func (h *MediaHandler) mediaResponse(c echo.Context, m models.Medium) map[string]interface{} {
	return mediaToResponse(m, h.mediaService.ThumbnailGeneration(c.Request().Context()))
}

func (h *MediaHandler) UploadFile(c echo.Context) error {
	file, err := c.FormFile("file")
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "file is required")
	}

	src, err := file.Open()
	if err != nil {
		return err
	}
	defer func() {
		_ = src.Close()
	}()

	content, err := io.ReadAll(src)
	if err != nil {
		return err
	}

	// Resolve the MIME type from the file's own magic bytes, ignoring the
	// client-supplied Content-Type. Rejects anything that isn't an allowlisted
	// image/video/audio format (e.g. an HTML page renamed to .jpg).
	mimeType, err := services.DetectMediaType(content, file.Header.Get("Content-Type"))
	if err != nil {
		return echo.NewHTTPError(http.StatusUnsupportedMediaType, uploadRejectionMessage(err))
	}

	altText := c.FormValue("alt_text")
	caption := c.FormValue("caption")
	postIDStr := c.FormValue("post_id")
	var postID *int64
	if postIDStr != "" {
		id, err := strconv.ParseInt(postIDStr, 10, 64)
		if err == nil {
			postID = &id
		}
	}

	media, err := h.mediaService.UploadFile(c.Request().Context(), services.UploadFileParams{
		Content:  content,
		Filename: file.Filename,
		MimeType: mimeType,
		AltText:  altText,
		Caption:  caption,
		PostID:   postID,
	})
	if err != nil {
		return MapError(err)
	}

	// A video upload may carry a poster frame the browser grabbed off the file
	// before sending it — the only way this build gets a video still, since the
	// server has no video decoder. Best-effort: a rejected or missing poster
	// leaves the video thumbnail-less rather than failing the upload.
	if poster, err := posterFromForm(c); err == nil && len(poster) > 0 {
		if updated, err := h.mediaService.SaveVideoPoster(c.Request().Context(), media.ID, poster); err == nil {
			media = updated
		}
	}

	return c.JSON(http.StatusCreated, h.mediaResponse(c, media))
}

// posterFromForm reads the optional "poster" upload field and returns its bytes
// once they have been confirmed to be an allowlisted image. It returns a nil
// slice and no error when the field is absent.
func posterFromForm(c echo.Context) ([]byte, error) {
	fh, err := c.FormFile("poster")
	if err != nil {
		return nil, nil
	}
	f, err := fh.Open()
	if err != nil {
		return nil, err
	}
	defer func() {
		_ = f.Close()
	}()

	// Cap the read: a poster is a single downscaled frame, and the bytes are
	// attacker-controlled up to whatever the upload limit allows.
	content, err := io.ReadAll(io.LimitReader(f, maxPosterBytes))
	if err != nil {
		return nil, err
	}

	mimeType, err := services.DetectMediaType(content, fh.Header.Get("Content-Type"))
	if err != nil {
		return nil, err
	}
	if !strings.HasPrefix(mimeType, "image/") {
		return nil, services.ErrUnsupportedMediaType
	}
	return content, nil
}

// SetVideoPoster stores a client-captured frame as an existing video's
// thumbnail. It backfills videos that arrived without one — anything uploaded
// before posters existed, or ingested over the API/MCP rather than the admin UI.
func (h *MediaHandler) SetVideoPoster(c echo.Context) error {
	id, err := parseIDParam(c)
	if err != nil {
		return err
	}

	poster, err := posterFromForm(c)
	if err != nil {
		return echo.NewHTTPError(http.StatusUnsupportedMediaType, uploadRejectionMessage(err))
	}
	if len(poster) == 0 {
		return echo.NewHTTPError(http.StatusBadRequest, "poster is required")
	}

	media, err := h.mediaService.SaveVideoPoster(c.Request().Context(), id, poster)
	if err != nil {
		return MapError(err)
	}
	return c.JSON(http.StatusOK, h.mediaResponse(c, media))
}

func (h *MediaHandler) ListMedia(c echo.Context) error {
	page, perPage := ParsePaginationParams(c, 20)
	fileType := c.QueryParam("file_type")
	folder := c.QueryParam("folder")

	media, total, err := h.mediaService.ListMedia(c.Request().Context(), services.ListMediaParams{
		Page:     page,
		PerPage:  perPage,
		FileType: fileType,
		Folder:   folder,
	})
	if err != nil {
		return MapError(err)
	}

	pages := int(math.Ceil(float64(total) / float64(perPage)))
	if pages == 0 {
		pages = 1
	}

	gen := h.mediaService.ThumbnailGeneration(c.Request().Context())
	items := make([]map[string]interface{}, len(media))
	for i, m := range media {
		items[i] = mediaToResponse(m, gen)
	}
	return c.JSON(http.StatusOK, map[string]interface{}{
		"media":    items,
		"total":    total,
		"page":     page,
		"per_page": perPage,
		"pages":    pages,
	})
}

func (h *MediaHandler) GetMediaFolders(c echo.Context) error {
	fileType := c.QueryParam("file_type")
	folders, err := h.mediaService.GetMediaFolders(c.Request().Context(), fileType)
	if err != nil {
		return MapError(err)
	}

	items := make([]map[string]interface{}, 0, len(folders))
	for _, f := range folders {
		items = append(items, map[string]interface{}{
			"year":  f.Year,
			"month": f.Month,
			"path":  f.Year + "/" + f.Month,
		})
	}
	return c.JSON(http.StatusOK, map[string]interface{}{"folders": items})
}

func (h *MediaHandler) GetMedia(c echo.Context) error {
	id, err := parseIDParam(c)
	if err != nil {
		return err
	}

	media, err := h.mediaService.GetMediaByID(c.Request().Context(), id)
	if err != nil {
		return MapError(err)
	}

	return c.JSON(http.StatusOK, h.mediaResponse(c, media))
}

type UpdateMediaRequest struct {
	AltText  string                  `json:"alt_text"`
	Caption  string                  `json:"caption"`
	PostID   *int64                  `json:"post_id"`
	Metadata *map[string]interface{} `json:"metadata"`
}

func (h *MediaHandler) UpdateMedia(c echo.Context) error {
	id, err := parseIDParam(c)
	if err != nil {
		return err
	}

	var req UpdateMediaRequest
	if err := c.Bind(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid request")
	}

	media, err := h.mediaService.UpdateMedia(c.Request().Context(), services.UpdateMediaParams{
		ID:       id,
		AltText:  req.AltText,
		Caption:  req.Caption,
		PostID:   req.PostID,
		Metadata: req.Metadata,
	})
	if err != nil {
		return MapError(err)
	}

	return c.JSON(http.StatusOK, h.mediaResponse(c, media))
}

func (h *MediaHandler) ReextractEXIF(c echo.Context) error {
	id, err := parseIDParam(c)
	if err != nil {
		return err
	}
	media, err := h.mediaService.ReextractEXIF(c.Request().Context(), id)
	if err != nil {
		return MapError(err)
	}
	return c.JSON(http.StatusOK, h.mediaResponse(c, media))
}

// UpdateEXIFRequest is the body for PUT /api/media/:id/exif.
// Keys are EXIF field names; values must contain only [a-zA-Z0-9 ].
type UpdateEXIFRequest map[string]string

func (h *MediaHandler) UpdateEXIF(c echo.Context) error {
	id, err := parseIDParam(c)
	if err != nil {
		return err
	}

	var req UpdateEXIFRequest
	if err := c.Bind(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid request body")
	}

	media, err := h.mediaService.UpdateEXIF(c.Request().Context(), services.UpdateEXIFParams{
		ID:     id,
		Fields: req,
	})
	if err != nil {
		return MapError(err)
	}
	return c.JSON(http.StatusOK, h.mediaResponse(c, media))
}

func (h *MediaHandler) RevertEXIF(c echo.Context) error {
	id, err := parseIDParam(c)
	if err != nil {
		return err
	}

	media, err := h.mediaService.RevertEXIF(c.Request().Context(), id)
	if err != nil {
		return MapError(err)
	}
	return c.JSON(http.StatusOK, h.mediaResponse(c, media))
}

func (h *MediaHandler) ListOrphanedMedia(c echo.Context) error {
	page, perPage := ParsePaginationParams(c, 20)

	media, total, err := h.mediaService.ListOrphanedMedia(c.Request().Context(), page, perPage)
	if err != nil {
		return MapError(err)
	}

	pages := int(math.Ceil(float64(total) / float64(perPage)))
	if pages == 0 {
		pages = 1
	}

	gen := h.mediaService.ThumbnailGeneration(c.Request().Context())
	orphaned := make([]map[string]interface{}, len(media))
	for i, m := range media {
		orphaned[i] = mediaToResponse(m, gen)
	}
	return c.JSON(http.StatusOK, map[string]interface{}{
		"media":    orphaned,
		"total":    total,
		"page":     page,
		"per_page": perPage,
		"pages":    pages,
	})
}

type BulkDeleteRequest struct {
	IDs []int64 `json:"ids"`
}

func (h *MediaHandler) BulkDeleteMedia(c echo.Context) error {
	var req BulkDeleteRequest
	if err := c.Bind(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid request")
	}

	if len(req.IDs) == 0 {
		return echo.NewHTTPError(http.StatusBadRequest, "no ids provided")
	}

	count, err := h.mediaService.BulkDeleteMedia(c.Request().Context(), req.IDs)
	if err != nil {
		return MapError(err)
	}

	return c.JSON(http.StatusOK, map[string]interface{}{
		"deleted": count,
		"message": fmt.Sprintf("Deleted %d media files", count),
	})
}

func (h *MediaHandler) DeleteMedia(c echo.Context) error {
	id, err := parseIDParam(c)
	if err != nil {
		return err
	}

	err = h.mediaService.DeleteMedia(c.Request().Context(), id)
	if err != nil {
		return MapError(err)
	}

	return c.NoContent(http.StatusNoContent)
}

func (h *MediaHandler) UploadMultiple(c echo.Context) error {
	form, err := c.MultipartForm()
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid multipart form")
	}
	files := form.File["files"]
	if len(files) == 0 {
		return echo.NewHTTPError(http.StatusBadRequest, "no files provided")
	}

	postIDStr := c.FormValue("post_id")
	var postID *int64
	if postIDStr != "" {
		if id, err := strconv.ParseInt(postIDStr, 10, 64); err == nil {
			postID = &id
		}
	}

	gen := h.mediaService.ThumbnailGeneration(c.Request().Context())
	var uploaded []interface{}
	var failed []interface{}

	for _, fh := range files {
		src, err := fh.Open()
		if err != nil {
			failed = append(failed, map[string]string{"filename": fh.Filename, "error": err.Error()})
			continue
		}
		content, err := io.ReadAll(src)
		_ = src.Close()
		if err != nil {
			failed = append(failed, map[string]string{"filename": fh.Filename, "error": err.Error()})
			continue
		}

		mimeType, err := services.DetectMediaType(content, fh.Header.Get("Content-Type"))
		if err != nil {
			failed = append(failed, map[string]string{"filename": fh.Filename, "error": uploadRejectionMessage(err)})
			continue
		}

		media, err := h.mediaService.UploadFile(c.Request().Context(), services.UploadFileParams{
			Content:  content,
			Filename: fh.Filename,
			MimeType: mimeType,
			PostID:   postID,
		})
		if err != nil {
			failed = append(failed, map[string]string{"filename": fh.Filename, "error": err.Error()})
			continue
		}
		uploaded = append(uploaded, mediaToResponse(media, gen))
	}

	return c.JSON(http.StatusCreated, map[string]interface{}{
		"uploaded":       uploaded,
		"failed":         failed,
		"total_uploaded": len(uploaded),
		"total_failed":   len(failed),
	})
}

func (h *MediaHandler) GetStorageStats(c echo.Context) error {
	stats, err := h.mediaService.GetStorageStats(c.Request().Context())
	if err != nil {
		return MapError(err)
	}
	return c.JSON(http.StatusOK, stats)
}

type RenameMediaRequest struct {
	NewFilename string `json:"new_filename"`
}

func (h *MediaHandler) RenameMedia(c echo.Context) error {
	id, err := parseIDParam(c)
	if err != nil {
		return err
	}

	var req RenameMediaRequest
	if err := c.Bind(&req); err != nil || req.NewFilename == "" {
		return echo.NewHTTPError(http.StatusBadRequest, "new_filename is required")
	}

	// Validate: only letters, digits, hyphens, underscores and spaces allowed in the base name.
	validName := regexp.MustCompile(`^[a-zA-Z0-9_\-\.\s]+$`)
	if !validName.MatchString(req.NewFilename) {
		return echo.NewHTTPError(http.StatusBadRequest, "filename may only contain letters, digits, hyphens, underscores and spaces")
	}

	media, err := h.mediaService.RenameMedia(c.Request().Context(), id, req.NewFilename)
	if err != nil {
		return MapError(err)
	}

	return c.JSON(http.StatusOK, h.mediaResponse(c, media))
}

func (h *MediaHandler) DeleteOrphanedMedia(c echo.Context) error {
	count, freed, err := h.mediaService.CleanupOrphaned(c.Request().Context())
	if err != nil {
		return MapError(err)
	}

	return c.JSON(http.StatusOK, map[string]interface{}{
		"message":       fmt.Sprintf("Deleted %d orphaned files", count),
		"deleted_count": count,
		"freed_bytes":   freed,
		"failed_count":  0,
	})
}

// RebuildThumbnails rolls the thumbnail generation token and purges the derived
// images. It takes no parameters: a rebuild is all-or-nothing now, and the old
// only_missing flag has nothing left to mean once every file is discarded.
func (h *MediaHandler) RebuildThumbnails(c echo.Context) error {
	res, err := h.mediaService.RebuildThumbnails(c.Request().Context())
	if err != nil {
		return MapError(err)
	}

	return c.JSON(http.StatusOK, map[string]interface{}{
		"message": fmt.Sprintf(
			"Thumbnails invalidated. Removed %d cached images; regenerating the %d most recent in the background.",
			res.Purged+res.Legacy, res.Prewarming),
		"stats": res,
	})
}

func (h *MediaHandler) AnalyzeImage(c echo.Context) error {
	file, err := c.FormFile("image")
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "image file is required")
	}

	src, err := file.Open()
	if err != nil {
		return err
	}
	defer func() {
		_ = src.Close()
	}()

	content, err := io.ReadAll(src)
	if err != nil {
		return err
	}

	analysis, err := h.mediaService.AnalyzeImage(c.Request().Context(), content, file.Filename, file.Header.Get("Content-Type"))
	if err != nil {
		return MapError(err)
	}

	return c.JSON(http.StatusOK, analysis)
}

func (h *MediaHandler) AnalyzeImageByPath(c echo.Context) error {
	var req struct {
		Path string `json:"path"`
	}
	if err := c.Bind(&req); err != nil || req.Path == "" {
		return echo.NewHTTPError(http.StatusBadRequest, "path is required")
	}

	analysis, err := h.mediaService.AnalyzeMediaByPath(c.Request().Context(), req.Path)
	if err != nil {
		return MapError(err)
	}

	return c.JSON(http.StatusOK, analysis)
}

func (h *MediaHandler) AnalyzeImageByID(c echo.Context) error {
	id, err := parseIDParam(c)
	if err != nil {
		return err
	}

	analysis, err := h.mediaService.AnalyzeMediaByID(c.Request().Context(), id)
	if err != nil {
		return MapError(err)
	}

	return c.JSON(http.StatusOK, analysis)
}

// uploadRejectionMessage turns a DetectMediaType failure into something an
// admin can act on. An SVG rejected for carrying a script is a different
// problem from a .exe renamed to .jpg, and "unsupported file type" describes
// only the second.
func uploadRejectionMessage(err error) string {
	if errors.Is(err, services.ErrActiveSVGContent) {
		return err.Error()
	}
	return "unsupported file type"
}
